mlflow/langchain/_langchain_autolog.py

```diff
@@ -85,35 +89,139 @@ def _invoke(self, *args, **kwargs):
+            return args, kwargs
+
+    elif func_name == "get_relevant_documents":
+        # callbacks is only available as kwargs in get_relevant_documents function
+        # https://github.com/langchain-ai/langchain/blob/7d444724d7582386de347fb928619c2243bd0e55/libs/core/langchain_core/retrievers.py#L173
+        callbacks = kwargs.get("callbacks") or []
+        kwargs["callbacks"] = _inject_callback(callbacks, mlflow_tracer)
+        return args, kwargs
+
     else:
-        yield args, kwargs
+        _logger.warning(f"Unsupported function `{func_name}`. Skipping injecting MLflow callbacks.")
+        return args, kwargs
+
+
+def _get_runnable_config_with_callback(
+    original_config: Union[None, RunnableConfig, List[RunnableConfig]],
+    new_callback: BaseCallbackHandler,
+) -> Union[RunnableConfig, List[RunnableConfig]]:
+    """
+    Create a new RunnableConfig (or a list of them) with the new callback injected.
+
+    This function MUST return a new RunnableConfig instance, instead of mutating the original
+    config. This is because the original config may be shared across different calls and in-place
+    modification may cause unexpected behaviors, e.g. double injection.
+
+    Args:
+        original_config: the original RunnableConfig passed by the user
+        new_callback: a new callback to be injected
+    """
+    if original_config is None:
+        return RunnableConfig(callbacks=[new_callback])
+    elif isinstance(original_config, list):
+        return [_get_runnable_config_with_callback(c, new_callback) for c in original_config]
+    elif isinstance(original_config, dict):
+        config_copy = original_config.copy()
+        callbacks = config_copy.pop("callbacks", None) or []
+        callbacks = _inject_callback(callbacks, new_callback)
+        return RunnableConfig(callbacks=callbacks, **config_copy)
+    else:
+        _logger.warning(
+            f"Unsupported config type `{original_config}` for autologging with tracing."
+            "Skipping injecting MLflow callbacks."
+        )
+        return original_config
+
+
+def _inject_callback(
+    original_callbacks: Union[List, BaseCallbackManager], new_callback: MlflowLangchainTracer
+) -> Union[List, BaseCallbackManager]:
+    """
+    Inject a callback into the original callbacks.
+
+    This function MUST return a new list or new callback manager instance, instead of
+    mutating the original callbacks. This is because the original callbacks may be shared across
+    different calls and in-place modification may cause unexpected behaviors, e.g. double injection.
+
+    Args:
+        original_callbacks: the original callbacks passed by the user
+        new_callback: a new callback to be injected
+    """
+    if isinstance(original_callbacks, BaseCallbackManager):
+        callback_manager_copy = original_callbacks.copy()
+        if not any(isinstance(cb, type(new_callback)) for cb in callback_manager_copy.handlers):
```

Comment:
While I thoroughly appreciate this approach of type checking of the `new_callback` arg, since we don't have actual enforcement of type submission for the usage of `new_callback`, should we explicitly require the type to be specified here to prevent regressions?

Reply:
Hmm this is not for checking the type of `new_calback`, but rather checking if it already exists in the callback list provided by the users. We have it in the current code, too.
https://github.com/mlflow/mlflow/blob/1effffdee0e635c5dfa55ab16422496590545e2e/mlflow/langchain/_langchain_autolog.py#L252-L256

Reply:
and that's what I get for scanning the code to quickly. Disregard as it's local type checking!